### ISSUE TITLE
Parallelize h5 operations in generate_stats_figs.R

### ISSUE THREAD

**Issue by ewallace** [riboviz/riboviz#106]
Change  h5 access in `generate_stats_figs.R` to be parallelizable, replacing the current situation where there is a single file handle repeatedly accessed by `rhdf5::H5Dopen` inside `GetGeneDatamatrix` etc.

We want to replace this by functions that take the file path as input (not file handle), so can parallize. Apparently the function `rhdf5::h5read` will do this.

Related to and required for  #32

**Comment by kavousan** [comment]
@FlicAnderson to invest half a day or so on this, see if it unblocks us (and Ania).

**Comment by FlicAnderson** [comment]
Created new branch `h5-open-106` from develop to resolve this issue, rather than from `regen-stats-32`, to avoid horrors/scope-creep in branch regen-stats-32.  

May either cherry pick commits from regen-stats-32 if single-issue, but if that's not easy/possible may just recreate changes in this separate branch for clarity of fixing and testing.

**Comment by FlicAnderson** [comment]
> Created new branch `h5-open-106` from develop to resolve this issue, rather than from `regen-stats-32`, to avoid horrors/scope-creep in branch regen-stats-32. 
> 
> May either cherry pick commits from regen-stats-32 if single-issue, but if that's not easy/possible may just recreate changes in this separate branch for clarity of fixing and testing.

Started with tentative fix process of this issue as applied in `regen-stats-32` branch, with 2 changes: 

In `generate_stats_figs.R`, **remove** `hdf5file <- rhdf5::H5Fopen(hd_file)` and **replace** with `hdf5file <- hd_file`
In `read_count_functions.R`, **change** function definition from this: 
```R
GetGeneDatamatrix <- function(gene, dataset, hdf5file) {
  hdf5file %>%
    rhdf5::H5Dopen(
      name = paste0("/", gene, "/", dataset, "/reads/data")
    ) %>%
    rhdf5::H5Dread() %>%
    return()
}
```
to this: 
```R
GetGeneDatamatrix <- function(gene, dataset, hdf5file){
  rhdf5::h5read(file = hdf5file, name = paste0("/", gene, "/", dataset, "/reads/data")) %>%
    return()
}
```
In branch `h5-open-106`, both changes are made in `generate_stats_figs.R` because `read_counts_functions.R` doesn't exist in develop branch, as it's a big addition from #32. 

Current vignette run status (commit 32cf983a): fails for each sample as gives error during `generate_stats_figs.R` stage for each sample. 

On interactive investigation: 
```R
[1] "Starting: Check for 3nt periodicity globally"
Saving 7 x 7 in image
Warning message:
In write.table(gene_pos_counts_bothends, file = tsv_file_path, append = T,  :
  appending column names to file
[1] "Completed: Check for 3nt periodicity globally"
[1] "Starting: Distribution of lengths of all mapped reads"
Error: Error in h5checktype(). Argument not of class H5IdComponent.
Execution halted
```
Error seems to lie in: 
```R
> gene_sp_read_length <- lapply(gene_names, function(x) {
+   rhdf5::H5Aread(rhdf5::H5Aopen(rhdf5::H5Gopen(hdf5file, paste0("/", x, "/", dataset, "/reads")), "reads_by_len"))
+ })
Error: Error in h5checktype(). Argument not of class H5IdComponent.
```
So this section equates to the function `GetGeneReadLength()` in `read_counts_functions.R` in branch `regen-stats-32`, and is because I haven't changed out the `rhdf5::H5Aread()`, `rhdf5::H5Aopen(), `rhdf5::H5Gopen()` functions yet. 

The same 3-function combo needs checking/updating in the following functions in `h5-open-106` branch in `generate_stats_figs.R` (apart from the bit of code above equivalent to `GetGeneReadLength()` as just mentioned): 
* `GetGeneLength()`
* `GetGeneReadsTotal()`

**These should help fix the problem, and is definitely my next step to fix this problem.**  Think I stopped the .h5 investigation halfway to swap tasks and missed these other functions when I went back to it.

These functions are quite easily accessible and fixable in `regen-stats-32` branch too, but are located in `read_counts_functions.R` rather than `generate_stats_figs.R`.

**Comment by FlicAnderson** [comment]
This is the time to start unit testing!

Suggestion, try out: {TestThat} library. 

Find a way to save the output and the inputs at key stages, take one or two examples of the input/output relationship, save them, then write test: assert that when we run the function then the output is equal to this. TestThat has tools to build on the assert but key points are, check known input and known output are what are expected despite changes to code.

Steps of suggested process:
- Define a test for the function
- Change the code to use the different method of accessing .h5 file
- Check test still passes
- If it does pass, then done.

Run this for all the data-getting/opening() functions, then you're done.

Useful technique for #32 refactoring.

**Comment by FlicAnderson** [comment]
Working on testing the .h5 open function changes in branch `h5-open-106`. 

Created temporary new script `test_h5-open-106.R` to do this, sourcing `generate_stats_figs.R`.

4 areas need to be fixed:

- GetGeneDatamatrix()
- gene_sp_read_length
- GetGeneLength()
- GetGeneReadsTotal()

I've been working on creating basic (ie non-TestThat) tests for now, checking structures, lengths and ensuring functions work in isolation when swapped out for alternatives which can take the .h5 file location instead of requiring an object of class `H5IdComponent` as H5Aread() or H5Gopen() and the various others currently in use do. 

### GetGeneDatamatrix()

As I mentioned [previously](https://github.com/riboviz/riboviz/issues/106#issuecomment-611003612), GetGeneDatamatrix() seems to be fixable quite easily by swapping out the functions and editing the line which assigns `hdf5file` a value, giving it the `hd_file` value, which is the location of the .h5 file for the sample.

```R
# OLD VERSION:
hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file

GetGeneDatamatrix <- function(gene, dataset, hdf5file) {
  hdf5file %>%
    rhdf5::H5Dopen(
      name = paste0("/", gene, "/", dataset, "/reads/data")
    ) %>%
    rhdf5::H5Dread() %>%
    return()
}

# NEW VERSION
hdf5file <- hd_file # filehandle for the h5 file

GetGeneDatamatrixAlt <- function(gene, dataset, hdf5file){
  rhdf5::h5read(file = hdf5file, name = paste0("/", gene, "/", dataset, "/reads/data")) %>%
    return()
}
```
Running this from the console for the WTnone sample using the following (after running the vignette process to create the .h5 files to have something to access):
```console
R --vanilla --args --num-processes=1 --min-read-length=10 --max-read-length=50 --buffer=250 --primary-id=Name --dataset=vignette --hd-file=vignette/output/WTnone/WTnone.h5 --orf-fasta-file=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True --output-dir=vignette/output/WTnone --do-pos-sp-nt-freq=True --t-rna-file=data/yeast_tRNAs.tsv --codon-positions-file=data/yeast_codon_pos_i200.RData --features-file=data/yeast_features.tsv --orf-gff-file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3 --asite-disp-length-file=data/yeast_standard_asite_disp_length.txt --count-threshold=64
```
Then within R, I've run the original and alternate/new version of the function, and am comparing outputs are equal:

```R
# old version outputs:
GeneDatamatrixYAL001C_OLD <- GetGeneDatamatrix("YAL001C", "vignette", hdf5file)

# run new version
GeneDatamatrixYAL001C_NEW <- GetGeneDatamatrixAlt("YAL001C", "vignette", hdf5file)

# check output structure
str(GeneDatamatrixYAL001C_NEW)
# int [1:41, 1:3983] 0 0 0 0 0 0 0 0 0 0 ...

# are outputs equal between old and new?
all.equal(GeneDatamatrixYAL001C_OLD, GeneDatamatrixYAL001C_NEW)
# [1] TRUE
```
... which they are. So far so good.

The next 3 areas which need fixing don't just involve opening a file connection, they involve pulling out and reading attributes within the .h5 file.  But because we are changing to functions which work from the .h5 file directly rather than via a separate connection object, these involve a bit more than just swapping out one function for another, particularly since these functions involve different nested H5 functions. 

I'm in the process of working out what each of those current read/open functions do, and replacing them with the appropriate alternative.  

### gene_sp_read_length

```R
# gene_sp_read_length:

# OLD VERSION CODE:
# hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file

# > gene_sp_read_length <- lapply(gene_names, function(x) {
#   +   rhdf5::H5Aread(rhdf5::H5Aopen(rhdf5::H5Gopen(hdf5file, paste0("/", x, "/", dataset, "/reads")), "reads_by_len"))
#   + })

# NEW VERSION CODE: 
# TODO

# RUN OLD VERSION: 
hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file

gene_sp_read_length <- lapply(gene_names, function(x) {
   rhdf5::H5Aread(rhdf5::H5Aopen(rhdf5::H5Gopen(hdf5file, paste0("/", x, "/", dataset, "/reads")), "reads_by_len"))
})

str(gene_sp_read_length)
# List of 68
# $ : num [1:41(1d)] 0 0 0 0 0 1 0 0 0 0 ...
# $ : num [1:41(1d)] 0 0 0 0 0 0 0 0 0 0 ...
# $ : num [1:41(1d)] 0 0 0 0 1 0 3 3 9 8 ...
# $ : num [1:41(1d)] 0 0 0 3 5 3 15 16 17 29 ...
# ...

length(gene_sp_read_length)

```
I still need to work on this, but the work I've started with GetGeneReadsTotal() (see below) should solve this pretty quickly since it's the same 3 functions nested. 

### GetGeneLength()
```R
# OLD VERSION CODE: 
# hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file
# # read length-specific total counts stored as attributes of 'reads_total' in H5 file
# GetGeneLength <- function(gene, hdf5file) {
#   start_codon_pos <- rhdf5::H5Aread(rhdf5::H5Aopen(rhdf5::H5Gopen(hdf5file, paste0("/", gene, "/", dataset, "/reads")), "start_codon_pos"))[1]
#   stop_codon_pos <- rhdf5::H5Aread(rhdf5::H5Aopen(rhdf5::H5Gopen(hdf5file, paste0("/", gene, "/", dataset, "/reads")), "stop_codon_pos"))[1]
#   return(stop_codon_pos - start_codon_pos)
# }

# NEW VERSION CODE: 
# TODO

# RUN OLD VERSION:
hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file
GetGeneLength <- function(gene, hdf5file) {
  start_codon_pos <- rhdf5::H5Aread(rhdf5::H5Aopen(rhdf5::H5Gopen(hdf5file, paste0("/", gene, "/", dataset, "/reads")), "start_codon_pos"))[1]
  stop_codon_pos <- rhdf5::H5Aread(rhdf5::H5Aopen(rhdf5::H5Gopen(hdf5file, paste0("/", gene, "/", dataset, "/reads")), "stop_codon_pos"))[1]
  return(stop_codon_pos - start_codon_pos)
}
GeneLengthYAL001C_OLD <- GetGeneLength("YAL001C", hdf5file)

# check output structure etc
str(GeneLengthYAL001C_OLD)
# int 3480

# RUN NEW VERSION CODE: 
# TODO

# COMPARE OLD AND NEW:
# TODO

```
Same deal; fixing GetGeneReadsTotal() (see below) should solve this pretty quickly since it's the same 3 functions nested.

### GetGeneReadsTotal()

```R


# OLD VERSION CODE:
# hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file
# GetGeneReadsTotal <- function(gene, hdf5file) {
#   rhdf5::H5Aread(rhdf5::H5Aopen(rhdf5::H5Gopen(hdf5file, paste0("/", gene, "/", dataset, "/reads")), "reads_total"))
# }

# NEW VERSION CODE:
# hdf5file <- hd_file # filehandle for the h5 file
# GetGeneReadsTotalAlt <- function(gene, hdf5file){
#   rhdf5::h5readAttributes(hd_file, name=paste0("/", "YAL001C", "/", "vignette", "/reads"))[5]
# }

# RUN OLD VERSION:
hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file
GetGeneReadsTotal <- function(gene, hdf5file) {
  rhdf5::H5Aread(rhdf5::H5Aopen(rhdf5::H5Gopen(hdf5file, paste0("/", gene, "/", dataset, "/reads")), "reads_total"))
}
GeneReadsTotalYAL001C_OLD <- GetGeneReadsTotal("YAL001C", hdf5file)

# check output structure etc
str(GeneReadsTotalYAL001C_OLD)
# num [1(1d)] 4
head(GeneReadsTotalYAL001C_OLD)
# [1] 4
length(GeneReadsTotalYAL001C_OLD)
#[1] 1


# RUN NEW VERSION CODE:
hdf5file <- hd_file # filehandle for the h5 file
GetGeneReadsTotalAlt <- function(gene, hdf5file){
  rhdf5::h5readAttributes(hd_file, name=paste0("/", "YAL001C", "/", "vignette", "/reads"))[5]
}
GeneReadsTotalYAL001C_NEW <- GetGeneReadsTotalAlt("YAL001C", hdf5file)

# check output structure etc
str(GeneReadsTotalYAL001C_NEW)
#List of 1
#$ reads_total: num [1(1d)] 4

head(GeneReadsTotalYAL001C_NEW)
# $reads_total
# [1] 4

length(GeneReadsTotalYAL001C_NEW)
# [1] 1

# are outputs equal between old and new?
all.equal(GeneReadsTotalYAL001C_OLD, GeneReadsTotalYAL001C_NEW)
# [1] "Modes: numeric, list"                                             
# [2] "names for current but not for target"                             
# [3] "Attributes: < names for target but not for current >"             
# [4] "Attributes: < Length mismatch: comparison on first 0 components >"
# [5] "target is array, current is list"  
```

With this one, I'm pretty close to getting it sorted. The outputs are different because my subsetting isn't quite right. _I need to access $reads_total and unlist it probably to match the current output, so this is my next move._

I've looked into the arguments of the current nested functions, and tried to understand them better so I can replicate the functionality

```console
# hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file
# GetGeneReadsTotal <- function(gene, hdf5file) {
#   rhdf5::H5Aread(rhdf5::H5Aopen(rhdf5::H5Gopen(hdf5file, paste0("/", gene, "/", dataset, "/reads")), "reads_total"))
# }

 ## H5Gopen(h5loc, name):
 # h5loc, h5 location identifier - file or group: hdf5file
 # name of group: paste0("/", gene, "/", dataset, "/reads")
 # opens groups of objects /gene/dataset/reads at h5 location hdf5file

 ## H5Aopen(h5obj, name):
 # h5obj: hdf5file, 
 # name of the attribute (character): "reads_total"
 # opens attribute "reads_total" from h5obj hdf5file, to be read by H5Aread

 ## H5Aread(h5attribute, buf=NULL):
 # h5attribute: the result of H5Aopen() on "reads_total" attribute
 # buf: a buffer related argument, not currently relevant
 # reads attribute object as created by H5Aopen
```

**Comment by FlicAnderson** [comment]
Following Edward's comment, will update the `gene_sp_read_length` bit with info from `regen-stats-32` branch to make it into a function (copy updated code across) to make this `h5-open-106` code up to date. 

Notes about this testing to stay in this issue tracker for now, but will be mined out into the TestThat testing framework later.

**Comment by FlicAnderson** [comment]
In branch `h5-open-106`, completed test script `rscripts/test_h5-open-106.R`, to test the 4 areas identified as needing fixes for improved .h5 access.

Following the inital exploration work done [here](https://github.com/riboviz/riboviz/issues/106#issuecomment-617831273), the original version of the code was added for comparison, and a new version created and tested for the same structure and results on usage. 

Each function was tested for 1 gene "YAL001C" from the "vignette" dataset, with hdf5file "vignette/output/WTnone/WTnone.h5", and code covering `gene_sp_read_length` was tested across  all genes as this code involves an `lapply()`. 

The outputs were saved and compared by eye for structure, and for equality using `all.equal()`.

The fixes all pass and seemed to substitute well for the original code, so these fixes were then encorporated into `generate_stats_figs.R` in commit [3d3c4f6](https://github.com/riboviz/riboviz/commit/3d3c4f650c9eb16c9f565eaa791d46ee7a40e141). 

This fix was then tested by running `generate_stats_figs.R` from terminal using the following :
```console
R --vanilla --args --num-processes=1 --min-read-length=10 --max-read-length=50 --buffer=250 --primary-id=Name --dataset=vignette --hd-file=vignette/output/WTnone/WTnone.h5 --orf-fasta-file=vignette/input/yeast_YAL_CDS_w_250utrs.fa --rpf=True --output-dir=vignette/output/WTnone --do-pos-sp-nt-freq=True --t-rna-file=data/yeast_tRNAs.tsv --codon-positions-file=data/yeast_codon_pos_i200.RData --features-file=data/yeast_features.tsv --orf-gff-file=vignette/input/yeast_YAL_CDS_w_250utrs.gff3 --asite-disp-length-file=data/yeast_standard_asite_disp_length.txt --count-threshold=64
```
which made the appropriate parameters available. 
I then sourced  `generate_stats_figs.R`:
```R
> source("/home/flic/riboviz/rscripts/generate_stats_figs.R")

Attaching package: ‘git2r’

The following objects are masked from ‘package:purrr’:

    is_empty, when

The following object is masked from ‘package:magrittr’:

    add

The following object is masked from ‘package:dplyr’:

    pull

The following object is masked from ‘package:rtracklayer’:

    as.data.frame

The following objects are masked from ‘package:Biostrings’:

    as.data.frame, head

The following object is masked from ‘package:XVector’:

    as.data.frame

The following objects are masked from ‘package:GenomicRanges’:

    as.data.frame, merge

The following objects are masked from ‘package:GenomeInfoDb’:

    as.data.frame, merge

The following objects are masked from ‘package:IRanges’:

    as.data.frame, diff, merge

The following objects are masked from ‘package:S4Vectors’:

    as.data.frame, head, merge

The following object is masked from ‘package:BiocGenerics’:

    as.data.frame

[1] "Created by: RiboViz"
[1] "Date: 2020-04-28 18:44:42"
[1] "File: rscripts/generate_stats_figs.R"
[1] "Version: commit 3d3c4f650c9eb16c9f565eaa791d46ee7a40e141 date 2020-04-28 17:44:11"
[1] "generate_stats_figs.R running with parameters:"
[1] "Starting: Check for 3nt periodicity globally"
Saving 7 x 7 in image
[1] "Completed: Check for 3nt periodicity globally"
[1] "Starting: Distribution of lengths of all mapped reads"
Saving 7 x 7 in image
[1] "Completed: Distribution of lengths of all mapped reads"
[1] "Starting: Biases in nucleotide composition along mapped read lengths"
[1] "Completed nucleotide composition bias table"
[1] "Starting: Check for 3nt periodicity (frame) by Gene"
Parsed with column specification:
cols(
  read_length = col_double(),
  asite_disp = col_double()
)
[1] "Completed: Check for 3nt periodicity (frame) by Gene"
[1] "Starting: Position specific distribution of reads"
Saving 7 x 7 in image
[1] "Completed: Position specific distribution of reads"
[1] "Starting: Calculate TPMs of genes"
[1] "Starting: Correlations between TPMs of genes with their sequence-based features"
Saving 7 x 7 in image
[1] "Completed: Correlations between TPMs of genes with their sequence-based features"
[1] "Starting: Codon-specific ribosome densities for correlations with tRNAs"
Saving 7 x 7 in image
[1] "Completed: Codon-specific ribosome densities for correlations with tRNAs"
There were 16 warnings (use warnings() to see them)
```
The warnings are fairly standard, and there were no errors on running.

I next tested that the vignette would run: 
```console
# vignette run: 
# 28 Apr 2020
# h5-open-106 branch, commit: 3d3c4f650c
flic@circe:~/riboviz$ source /home/flic/miniconda3/bin/activate
(base) flic@circe:~/riboviz$ source /home/flic/setenv.sh 
(base) flic@circe:~/riboviz$ time python -m riboviz.tools.prep_riboviz -c vignette/vignette_config.yaml
Running under Python: 3.7.3 (default, Mar 27 2019, 22:11:17) 
[GCC 7.3.0]
Created by: RiboViz Date: 2020-04-28 18:51:08.059236 Command-line tool: /home/flic/riboviz/riboviz/tools/prep_riboviz.py File: /home/flic/riboviz/riboviz/tools/prep_riboviz.py Version: commit 3d3c4f650c9eb16c9f565eaa791d46ee7a40e141 date 2020-04-28 18:44:11+01:00 
Configuration file: vignette/vignette_config.yaml
Command file: run_riboviz_vignette.sh
Number of processes: 1
# ...
Finished processing 3 samples, 1 failed
Collate TPMs across sample results. Log: vignette/logs/20200428-185108/collate_tpms.log
Count reads. Log: vignette/logs/20200428-185108/count_reads.log
Completed
real	2m19.132s
user	2m58.144s
sys	0m4.047s
```
This ran successfully (the 1 failed sample being the missing file which we expect to fail).

I then ran the regression test against the `regression-test-data-20200221` test data parcel from [here](https://github.com/riboviz/regression-test-data-20200221):
```console
# regression test against https://github.com/riboviz/regression-test-data-20200221 
(base) flic@circe:~/riboviz$ pytest -v riboviz/test/regression/test_vignette.py --expected=/home/flic/regression-test-data-20200221
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-5.0.1, py-1.8.0, pluggy-0.12.0 -- /home/flic/miniconda3/bin/python
cachedir: .pytest_cache
rootdir: /home/flic/riboviz
plugins: cov-2.8.1
collected 72 items                                                             
riboviz/test/regression/test_vignette.py::test_index[1-YAL_CDS_w_250] SKIPPED [  1%]
# ...
riboviz/test/regression/test_vignette.py::test_sample_output_pdf[3ntframe_propbygene.pdf-WTnone] PASSED [ 97%]
riboviz/test/regression/test_vignette.py::test_output_tsv[TPMs_collated.tsv] PASSED [ 98%]
riboviz/test/regression/test_vignette.py::test_output_tsv[read_counts.tsv] PASSED [100%]
=================== 42 passed, 30 skipped in 140.26 seconds ====================
```
Note: 30 are skipped as the test parcel does not include the index and tmp files.

All tested files pass the regression test, and all outputs for the vignette look reasonable and normal.

I'm therefore classing this as a success and submitting pull request for review!

**Comment by FlicAnderson** [comment]
Re-ran regression test with `--check-index-tmp`, regression test files created using current develop.
```console
(base) flic@circe:~/riboviz$ pytest -v riboviz/test/regression/test_vignette.py     --expected=$HOME/regression-test-data-develop-63241663f2 --check-index-tmp
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-5.0.1, py-1.8.0, pluggy-0.12.0 -- /home/flic/miniconda3/bin/python
cachedir: .pytest_cache
rootdir: /home/flic/riboviz
plugins: cov-2.8.1
collected 72 items                                                             
riboviz/test/regression/test_vignette.py::test_index[1-YAL_CDS_w_250] PASSED [  1%]
riboviz/test/regression/test_vignette.py::test_index[1-yeast_rRNA] PASSED [  2%]
# ... 
riboviz/test/regression/test_vignette.py::test_output_tsv[TPMs_collated.tsv] PASSED [ 98%]
riboviz/test/regression/test_vignette.py::test_output_tsv[read_counts.tsv] PASSED [100%]
========================= 72 passed in 222.70 seconds ==========================
```

Ran all tests (excluding regression) and including coverage test: 
```console
(base) flic@circe:~/riboviz$ pytest --cov-config=.coveragerc --cov-report term-missing --cov=riboviz --ignore-glob="*regression*"
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-5.0.1, py-1.8.0, pluggy-0.12.0
rootdir: /home/flic/riboviz
plugins: cov-2.8.1
collected 165 items                                                            

riboviz/test/test_barcodes_umis.py .................                     [ 10%]
riboviz/test/test_demultiplex_fastq.py ..................                [ 21%]
riboviz/test/test_fastq.py ........................                      [ 35%]
riboviz/test/test_process_utils.py ........................              [ 50%]
riboviz/test/test_sam_bam.py ................                            [ 60%]
riboviz/test/test_trim_5p_mismatch.py ............                       [ 67%]
riboviz/test/test_upgrade_config.py .....                                [ 70%]
riboviz/test/test_utils.py ....................                          [ 82%]
riboviz/test/tools/test_prep_riboviz.py ............                     [ 89%]
riboviz/test/tools/test_prep_riboviz_simdata_multiplex.py .............  [ 97%]
riboviz/test/tools/test_prep_riboviz_simdata_umi.py ....                 [100%]

----------- coverage: platform linux, python 3.7.3-final-0 -----------
Name                                    Stmts   Miss  Cover   Missing
---------------------------------------------------------------------
riboviz/__init__.py                         6      0   100%
riboviz/barcodes_umis.py                   25      0   100%
riboviz/bedgraph.py                        19     19     0%   4-72
riboviz/check_fasta_gff.py                 24     24     0%   4-58
riboviz/compare_files.py                   33     33     0%   4-62
riboviz/count_reads.py                    192     39    80%   133-135, 175-177, 222, 246-248, 251-264, 297-299, 334-336, 384-386, 389-394, 436, 441-443
riboviz/create_fastq_simdata.py           125    125     0%   77-474
riboviz/demultiplex_fastq.py              100     14    86%   233, 238-239, 247-261, 283, 289, 307, 316-319
riboviz/fastq.py                           48      0   100%
riboviz/h5.py                               7      3    57%   26-28
riboviz/hisat2.py                           3      3     0%   5-8
riboviz/logging_utils.py                   24      2    92%   75, 81
riboviz/params.py                          41      0   100%
riboviz/process_utils.py                   50      0   100%
riboviz/provenance.py                      31      2    94%   35-36
riboviz/sam_bam.py                        120     85    29%   101-124, 146-153, 167-179, 196-204, 223-230, 245, 260-346
riboviz/sample_sheets.py                   33      0   100%
riboviz/subsample_bioseqfile.py            23     23     0%   6-52
riboviz/tools/__init__.py                   0      0   100%
riboviz/tools/check_fasta_gff.py           18     18     0%   2-57
riboviz/tools/compare_files.py             18     18     0%   2-73
riboviz/tools/count_reads.py               24      0   100%
riboviz/tools/create_barcode_pairs.py      16     16     0%   2-59
riboviz/tools/create_fastq_simdata.py      14     14     0%   2-51
riboviz/tools/demultiplex_fastq.py         27      0   100%
riboviz/tools/prep_riboviz.py             281     32    89%   354-355, 367, 519-522, 617, 625, 692, 718, 751-759, 773, 819-820, 834-835, 851-864, 876-880, 884
riboviz/tools/subsample_bioseqfile.py      24     24     0%   2-105
riboviz/tools/trim_5p_mismatch.py          27      0   100%
riboviz/tools/upgrade_config_file.py       16     16     0%   2-60
riboviz/trim_5p_mismatch.py                80      7    91%   76-83, 129-132
riboviz/umi_tools.py                        5      0   100%
riboviz/upgrade_config.py                  35      1    97%   191
riboviz/utils.py                           45     13    71%   100-106, 121-123, 198-203
riboviz/workflow.py                       120      1    99%   619
riboviz/workflow_files.py                  20      0   100%
riboviz/workflow_r.py                      20      0   100%
---------------------------------------------------------------------
TOTAL                                    1694    532    69%

========================= 165 passed in 149.97 seconds =========================
```
This seems good; pull request submitted (#170 )

**Comment by kavousan** [comment]
To close together with pull request.

**Comment by FlicAnderson** [comment]
Merged in pull request #170. 

Note to dev team: requires installing package `here` in R via:
```R
install.packages("here")
```

**Comment by FlicAnderson** [comment]
> Merged in pull request #170.
> 
> Note to dev team: requires installing package `here` in R via:
> 
> ```r
> install.packages("here")
> ```

Realised I can notify folks of the merge and newly required R package using @riboviz/developers, but would emailing be preferable for something like this?
